File: h_w_s_w_interface/Practicals/data_representation/c/show_bytes.c
/*
 * show_bytes.c
 * ------------------------------------------
 * Demonstrates how different data types are represented
 * at the byte level in memory using C.
 *
 * Concept:
 *  - Any data type can be interpreted as a sequence of bytes.
 *  - By casting a variable’s address to (char*), we can examine
 *    each byte in memory.
 *  - This reveals details such as byte order (endianness) and
 *    how data types occupy memory.
 *
 * Compile and run:
 *   $ gcc -o show_bytes_c show_bytes.c
 *   $ ./show_bytes_c
 *
 * Or use the Makefile:
 *   $ make run-c
 */

#include <stdio.h>

/*
 * show_bytes()
 * -------------
 * Prints the byte-level representation of any data value.
 *
 * Parameters:
 *   start - pointer to the beginning of the data in memory
 *   len   - number of bytes to print
 *
 * Output format:
 *   Address    Byte Value (in hex)
 *
 * Example:
 *   0x7fff6f330dcc   0x39
 */

void show_bytes(char *start, int len)
{
    for (int i = 0; i < len; i++)
    {
        printf("%p\t0x%.2x\n", start + i, *(unsigned char *)(start + i));
    }
    printf("\n");
}

/*
 * Convenience wrappers for specific data types
 * ---------------------------------------------
 * Each function calls show_bytes() with the appropriate type and size.
 */

void show_int(int x)
{
    show_bytes((char *)&x, sizeof(int));
}

void show_float(float x)
{
    show_bytes((char *)&x, sizeof(float));
}

void show_double(double x)
{
    show_bytes((char *)&x, sizeof(double));
}

int main()
{
    int a = 12345;
    float b = 12345.0;
    double c = 12345.0;

    printf("int a = %d;\n", a);
    show_int(a);

    printf("float b = %.2f;\n", b);
    show_float(b);

    printf("double c = %.2f;\n", c);
    show_double(c);

    return 0;
}

File: h_w_s_w_interface/Practicals/data_representation/go/show_bytes.go
// Concept:
// - In Go, we can view values as byte slices using the
// encoding/binary package and the math package for float bits.
// - Like C, this reveals endianness and biray encoding details
// $ make run-go

package main

import (
	"encoding/binary"
	"fmt"
	"math"
	"unsafe"
)

// showBytes prints the memory address and value of each byte
// in the provided byte slice

func showBytes(data []byte) {
	for i, b := range data {
		fmt.Printf("%p\t0x%.2x\n", unsafe.Pointer(&data[i]), b)
	}
	fmt.Println()
}

// showInt demonstrates byte-level representation of an int32 value
func showInt(x int32) {
	fmt.Printf("int a = %d;\n", x)

	// convert int to byte array (Little Endian)
	buf := make([]byte, 4)
	binary.LittleEndian.PutUint32(buf, uint32(x))

	showBytes(buf)
}

// showFloat demostrates byte-level representation of a float32 value
func showFloat(x float32) {
	fmt.Printf("float b = %.2f;\n", x)

	// convert float to IEEE 754 bit representation
	buf := make([]byte, 4)
	binary.LittleEndian.PutUint32(buf, math.Float32bits(x))

	showBytes(buf)
}

// showDouble demonstrates byte-level representation of a float64 value
func showDouble(x float64) {
	fmt.Printf("double c = %.2f;\n", x)

	buf := make([]byte, 8)
	binary.LittleEndian.PutUint64(buf, math.Float64bits(x))

	showBytes(buf)
}

func main() {
	showInt(12345)
	showFloat(12345.0)
	showDouble(12345.0)
}
